pkg/tsdb/cloudwatch/live.go

```diff
@@ -0,0 +1,269 @@
+package cloudwatch
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
+	"github.com/centrifugal/centrifuge"
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/tsdb"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/sync/errgroup"
+)
+
+type LogQueryRunnerSupplier struct {
+}
+
+type LogQueryRunner struct {
+	channelName string
+	publish     models.ChannelPublisher
+	running     map[string]bool
+}
+
+const (
+	maxAttempts         = 8
+	initialPollInterval = 500 * time.Millisecond
+)
+
+var (
+	channelMu        = sync.Mutex{}
+	responseChannels = map[string]chan *tsdb.Response{}
```

Comment:
Can we avoid adding global variables?

Reply:
Please make these into struct fields, so we can avoid globals.

Reply:
To briefly outline the workflow:
A client makes an HTTP request with the query, and query type `liveLogAction`. The `cloudWatchExecutor` `Query` method is called, which then calls `executeLiveLogQuery`. From there, a UUID is generated, and a corresponding go channel is made, through which query responses will be sent and consumed. A response to the client's HTTP request is made which includes that UUID. At the same time, `StartQuery` requests are made to AWS, followed by periodic `GetQueryResults` requests, the results of which are sent to the go channel previously created.
When the client subscribes to the centrifuge channel, `OnSubscribe` is called and the responses from the corresponding go channel are taken and sent through the centrifuge channel back to the client until the go channel closes (i.e. the query has completed).

The issue is, there doesn't seem to be a simple way (ignoring global variables) to let `LogQueryRunner` have access to `cloudWatchExecutor`, since `LogQueryRunner` isn't created until a client subscribes and `GetHandlerForPath` is called. Furthermore there doesn't seem to be an easy way to access the executor - earlier I thought maybe since it's set in a registry using `RegisterTsdbQueryEndpoint` there might be a corresponding function to retrieve it, which there is (`getTsdbQueryEndpointFor`) but it isn't exposed outside the `tsdb` package.
I think it would make things easier if that function _were_ exposed, but perhaps there's a reason it's not. Any thoughts?

Reply:
Thanks @kaydelaney. It's too late for me tonight, but I can look more at this tomorrow.